Add exception raising for validation errors

RandyRcw opened the pull request:
Work in progress for issue #10733 

Added exception raising for validation errors and warnings and replaced env variable BOKEH_STRICT with BOKEH_VALIDATION_EXCEPTIONS

In a comment on the issue page (https://github.com/bokeh/bokeh/issues/10733#issuecomment-735002152), it was initially recommended for check_integrity to only be for returning messages, and to punt the potential exception raising downstream to the caller to avoid a muddled API. 
However, after changing BOKEH_STRICT to  BOKEH_VALIDATION EXCEPTIONS as recommended in this comment (https://github.com/bokeh/bokeh/issues/10733#issuecomment-735003520), I noticed that if potential exception raising was left in the caller, I wouldn't be able to access messages to check for a warning or error. In the original pull request made for this issue I saw a comment (https://github.com/bokeh/bokeh/pull/10780#r550380992) that said everything should be handled inside check_integrity, and so I was wondering if I should be doing potential exception raising in check_integrity? If not, how should I go about accessing messages from the validate function in document.py?

bryevdv commented on bokeh/document/document.py:
```diff
@@ -971,6 +971,10 @@ def validate(self):
             check_integrity(refs)
 
+            # Currently not sure how to access messages inside this function
+            if len(messages['error']) or (len(messages['warning']) and settings.strict()):
```
`settings.strict()` corresponds to `BOKEH_STRICT` this needs to be updated to process the new `settings.validation_exceptions()` instead

bryevdv replied:
But see note below about changing this name to `validation_level`

bryevdv commented on bokeh/settings.py:
```diff
@@ -608,8 +608,15 @@ def dev(self):
 
-    strict = PrioritizedSetting("strict", "BOKEH_STRICT", convert=convert_bool, help="""
-    Whether validation checks should be strict (i.e. raise errors).
+    validation_exceptions = PrioritizedSetting("validation_exceptions", "BOKEH_VALIDATION_EXCEPTIONS", default="none", help="""
```
After a little more thought, I think a better name would be `validation_level` / `BOKEH_VALIDATION_LEVEL`

bryevdv commented on bokeh/core/validation/check.py:
```diff
@@ -125,10 +125,6 @@ def check_integrity(models):
         if code not in __silencers__:
             log.warning("W-%d (%s): %s: %s" % msg)
```
All the `log` calls in this function will have to be removed. Instead, the function should  collect all the warnings and errors into some data structure, and then return that data structure.

bryevdv commented on bokeh/document/document.py:
```diff
@@ -971,6 +971,10 @@ def validate(self):
             refs = r.references()
             check_integrity(refs)
```
The new return value return value from `check_integrity` (discussed above) will need to be stored, and then processed below according to the validation level policy.

bryevdv replied:
Given the need for the processing to *also* take in to account any `__silencers__`, the best approach is probably to encapsulate the logic for emitting warnins into a new separate function in `bokeh.validatation`. The line above should more or less then end up like:
```
validations = check_integrity(refs)

process_validations(validations)
```
which will also help prevent cluttering up `document` more.

RandyRcw commented:
Thanks for the quick response and help, we've applied the changes you recommended. We were also wondering where the mocked tests for check_validate that you mentioned in issue #10733 (https://github.com/bokeh/bokeh/issues/10733#issuecomment-735002152) might exist?

bryevdv commented on bokeh/core/validation/check.py:
```diff
@@ -85,24 +85,24 @@ def silenced(warning: int) -> None:
 
         >>> check_integrity([empty_row])
-        W-1002 (EMPTY_LAYOUT): Layout has no children: Row(id='2404a029-c69b-4e30-9b7d-4b7b6cdaad5b', ...)
+        dict()
```
It would be nice to give a real (abridged) example output here to show the structure of the dict.

bryevdv commented on bokeh/document/document.py:
```diff
@@ -969,7 +970,16 @@ def validate(self):
+                    raise RuntimeError("Errors encountered during validation (see log output)")
+            elif settings.validation_level() == "all":
+                if len(validations['error']) or len(validations['warning']):
+                    raise RuntimeError("Errors encountered during validation (see log output)")
```
I would move all of this block to `process_validations`

bryevdv replied:
Also after some thought, maybe `issues` is a better name:
```
issues = check_integrity(refs)
process_validation_issuess(issues)
```

bryevdv commented on bokeh/settings.py:
```diff
@@ -608,8 +608,15 @@ def dev(self):
 
-    strict = PrioritizedSetting("strict", "BOKEH_STRICT", convert=convert_bool, help="""
-    Whether validation checks should be strict (i.e. raise errors).
+    validation_level = PrioritizedSetting("validation_level", "BOKEH_VALIDATION_LEVEL", default="none",help="""
```
We will want to make sure to give suitable error message if bad values are passed. This setting should model itself after the log level setting. See the `convert_logging` function here:

https://github.com/bokeh/bokeh/blob/bee3da0fbd4ec8b59b49f449ade9f7282481bb99/bokeh/settings.py#L212

RandyRcw commented:
Updated the process_validation function to handle the log warnings/errors, and added a convert_validation function to output an error message for bad values. 

We were also wondering how to properly set the value of an environment variable like BOKEH_VALIDATION_LEVEL? We've tried setting it in our tests with  
( BOKEH_VALIDATION_LEVEL="errors" ) and ( settings.validation_level("errors") ), but neither seems to actually change the value of the variable. 

Thanks again for the feedback so far!

bryevdv commented:
Assuming this is OSX or Linux, what is the output of 
```
BOKEH_VALIDATION_LEVEL="errors" python -c "from bokeh.settings import settings; print(settings.validation_level())"
```
?

bryevdv commented on bokeh/settings.py:
```diff
@@ -243,6 +243,27 @@ def convert_logging(value):
+        ValueError
+
+    '''
+    if value.lower() == "none" or value.lower() == "errors" or value.lower() == "all":
+        return value.lower()
+
+    raise ValueError("Cannot convert {} to validation level string, valid values are: none, errors, all".format(value))
```
There is no need to call `value.lower()` four times, also the list of valid values can be defined and re-used:
```suggestion
    VALID_LEVELS = ["none", "errors", "all"]
    
    value = value.lower()
    if value in VALID_LEVELS:
        return value

    raise ValueError(f"Cannot convert {value} to validation level string, valid values are: {VALID_LEVELS!r}")
```

bryevdv commented on bokeh/core/validation/check.py:
```diff
@@ -85,24 +88,24 @@ def silenced(warning: int) -> None:
 
         >>> check_integrity([empty_row])
-        W-1002 (EMPTY_LAYOUT): Layout has no children: Row(id='2404a029-c69b-4e30-9b7d-4b7b6cdaad5b', ...)
+        dict(warning = [(1002, EMPTY_LAYOUT, Layout has no children, Row(id='2404a029-c69b-4e30-9b7d-4b7b6cdaad5b', ...))])
```
I think this would, read better in pretty-printed "literal" form
```suggestion
        {
            "warning": [
                (1002, EMPTY_LAYOUT, Layout has no children, Row(id='2404a029-c69b-4e30-9b7d-4b7b6cdaad5b', ...),
                ...
            ],
            "error": [...]
        }             
```

RandyRcw commented:
Updated the logic for convert_validation and the docstring for check_integrity. I also modified the unit tests involving check_integrity to include process_validation_issues to make those unit tests functional again. 

Are there any other tests that I should be looking to add or modify to accommodate the changes made so far?

bryevdv commented on tests/unit/bokeh/test_settings.py:
```diff
@@ -90,11 +90,12 @@ def test_types(self) -> None:
 
         assert bs.settings.py_log_level.convert_type == "Log Level"
 
+        assert bs.settings.validation_level.convert_type == "Validation Level String"
```
Let's make this "Validation Level", to be consistent with "Log Level"

```suggestion
        assert bs.settings.validation_level.convert_type == "Validation Level"
```

bryevdv commented on bokeh/settings.py:
```diff
@@ -401,6 +425,7 @@ def convert_type(self):
         if self._convert is convert_bool: return "Bool"
         if self._convert is convert_logging: return "Log Level"
         if self._convert is convert_str_seq: return "List[String]"
+        if self._convert is convert_validation: return "Validation Level String"
```
```suggestion
        if self._convert is convert_validation: return "Validation Level"
```

bryevdv commented on bokeh/core/validation/check.py:
```diff
@@ -117,17 +126,40 @@ def check_integrity(models):
+            raise RuntimeError("Errors encountered during validation (see log output)")
+    elif settings.validation_level() == "all":
+        if len(issues['error']) or len(issues['warning']):
+            raise RuntimeError("Errors encountered during validation (see log output)")
```
We should not just say "see log output", the Runtime error should itself should be formatted to include all the warnings and/or errors as appropriate.

bryevdv replied:
I might suggest a code structure more like this:
```
# filter things first so it's not an issue later
errors = issues['errors']
warnings = [item for item in issues['warnings'] if item[0] not in __silencers__

# collect all the message that will get logged as they are logged
warning_messages = [] 
for code, name, desc, obj in sorted(warnings):
    msg = f"W-{code} ({name}): {desc}: {obj}"
    warning_messages.append(msg)
    log.warning(msg)

# do similar with errors

# raise exceptions, using warning_messages and errror_messages
```

bryevdv replied:
Also as a remindedr we always use f-strings, and do not use old-style string formatting anywhere.

bryevdv commented:
> Are there any other tests that I should be looking to add or modify to accommodate the changes made so far?

@RandyRcw You have updated the exisiting tests in `test_validation` to just use the two new functions wherever the single old one was used:
```
    issues = v.check_integrity([m])
    v.process_validation_issues(issues)
```
But it would be more in the spirit of unit testing to add tests that check the two new functions in isolation.

bryevdv commented on bokeh/core/validation/check.py:
```diff
@@ -117,17 +141,48 @@ def check_integrity(models):
+
+    error_messages = []
+    for code, name, desc, obj in sorted(errors):
+        msg = f"W-{code} ({name}): {desc}: {obj}"
```
```suggestion
        msg = f"E-{code} ({name}): {desc}: {obj}"
```

bryevdv commented on bokeh/core/validation/check.py:
```diff
@@ -117,17 +141,48 @@ def check_integrity(models):
+            raise RuntimeError(f"Errors encountered during validation: {error_messages}")
+    elif settings.validation_level() == "all":
+        if len(errors) or len(warnings):
+            raise RuntimeError(f"Errors encountered during validation: {error_messages} {warning_messages}")
```
```suggestion
            raise RuntimeError(f"Errors encountered during validation: {error_messages+warning_messages}")
```

bryevdv commented on bokeh/core/validation/check.py:
```diff
@@ -76,6 +79,21 @@ def silence(warning: int, silence: bool = True) -> Set[int]:
+        bool
+
+    '''
+    if len(warning) > 0:
+        code = warning[0]
+        return code in __silencers__
+    return False
```
I think this can just be:
```suggestion
    return code[0] in __silencers__
```
if warning is not a 4-tuple then something bad has gone wrong. 

The reason it's nice to define this predicate is that I think in  the future we may want to define a proper data structure, or at least a `namedtuple` so defining this predicate now will insulate against those changes.

bryevdv commented:
@RandyRcw A few small last comments, I think this is very close. Thanks for your patience!

RandyRcw commented:
@bryevdv That's great to hear! The most recent commit contains the changes made off of your feedback.

RandyRcw commented:
@bryevdv I fixed a couple of minor issues that were causing the CI tests to not pass. I think everything should be set now!

bryevdv commented on bokeh/settings.py:
```diff
@@ -96,7 +96,7 @@
 API
 ~~~
-
+d
```
This seems like a mistaken add?

RandyRcw replied:
Not sure how that got in there, but it should be good now.

bryevdv commented:
Just FYI an external update to `black` has broken our CI today. So this will need a merge/rebase on latest `branch-2.4` once that fix is in.

bryevdv commented on tests/unit/bokeh/test_settings.py:
```diff
@@ -158,6 +159,10 @@ def test_convert_logging_bad(self) -> None:
 
+    @pytest.mark.parametrize("value", ["none", "errors", "all"])
+    def test_convert_validation_good(self, value) -> None:
+        assert bs.convert_validation(value) == value
```
I missed that there was not a `test_convert_validation_bad` added. e.g. something like
```
    def test_convert_validation_bad(self) -> None:
        with pytest.raises(ValueError):
            bs.convert_validation("junk")
```
It's always important to probe the space of inputs to include bad values, to make sure that the expected behavior in those cases is intended, documented, and repeatable.

RandyRcw replied:
Thanks for catching that! The test for that has been added to test_settings.py now.

bryevdv commented:
The fix for the CI issue has been merged, so a merge/rebase on current `branch-2.4` should fix the failing codebase test. Let me know if you have questions about that.

RandyRcw commented:
@bryevdv Okay so I would merge the current branch-2.4 into my fix-issue-10733 branch to fix the codebase test?

bryevdv commented:
>  Okay so I would merge the current branch-2.4 into my fix-issue-10733 branch to fix the codebase test?

yes, or rebase on it and force push (either way, after pulling the latest `branch-2.4` from the main repo first)

RandyRcw commented:
@bryevdv Okay sounds good! So is the pull request fully ready to be merged now?

bryevdv commented:
Yes, will merge when the last three test pass in a few minutes (it looks like you must have picked up the branch-2.4 change already at some point)

bryevdv commented:
Thanks @RandyRcw !

RandyRcw commented:
@bryevdv Awesome! Thanks so much for all the help Bryan! My partner @danie1zhang and I found this to be a great experience and we were really able to improve as programmers because of your guidance. We appreciate how you were so quick to answer our questions and provide feedback for us. Bokeh is a really interesting project and we enjoyed working on it!
